pkg/storage/ingest/reader.go

```diff
@@ -169,11 +170,28 @@ func (r *PartitionReader) consumeFetches(ctx context.Context, fetches kgo.Fetche
 		})
 	})
 
-	err := r.consumer.consume(ctx, records)
-	if err != nil {
-		level.Error(r.logger).Log("msg", "encountered error processing records; skipping", "min_offset", minOffset, "max_offset", maxOffset, "err", err)
-		// TODO abort ingesting & back off if it's a server error, ignore error if it's a client error
+	boff := backoff.New(ctx, backoff.Config{
+		MinBackoff: 250 * time.Millisecond,
+		MaxBackoff: 2 * time.Second,
+		MaxRetries: 0, // retry forever
+	})
+
+	for boff.Ongoing() {
+		err := r.consumer.consume(ctx, records)
+		if err != nil {
+			level.Error(r.logger).Log(
+				"msg", "encountered error while consuming; will retry",
+				"err", err,
+				"record_min_offset", minOffset,
+				"record_max_offset", maxOffset,
+				"num_retries", boff.NumRetries(),
+			)
+			boff.Wait()
+			continue
+		}
+		break
```

Comment:
[nit] Code would be easier to read if above you do `if err == nil { break }`. This way you also don't need `continue` above.